subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4935,4 +5003,184 @@ static int lwm2m_engine_init(const struct device *dev)
+	/* did not read anything even if we should have - on single item */
+	if (ret == 0 && num_read == 0U && msg->path.level == 3U) {
+		return -ENOENT;
+	}
+
+	return ret;
+}
+
+static int do_send_op(struct lwm2m_message *msg, uint16_t content_format,
+		      struct lwm2m_obj_path path_list[], uint8_t path_list_size)
+{
+	switch (content_format) {
+#if defined(CONFIG_LWM2M_RW_SENML_JSON_SUPPORT)
+	case LWM2M_FORMAT_APP_SEML_JSON:
+		return do_send_op_senml_json(msg, path_list, path_list_size);
+#endif
+
+	default:
+		LOG_ERR("Unsupported content-format for /dp: %u", content_format);
+		return -ENOMSG;
+	}
+}
+
+static int do_send_reply_cb(const struct coap_packet *response,
+				 struct coap_reply *reply,
+				 const struct sockaddr *from)
+{
+	uint8_t code;
+
+	code = coap_header_get_code(response);
+	LOG_DBG("Send callback (code:%u.%u)",
+		COAP_RESPONSE_CODE_CLASS(code),
+		COAP_RESPONSE_CODE_DETAIL(code));
+
+	if (code == COAP_RESPONSE_CODE_CHANGED) {
+		LOG_INF("Send done!");
+		return 0;
+	}
+
+	LOG_ERR("Failed with code %u.%u. Not Retrying.",
+		COAP_RESPONSE_CODE_CLASS(code), COAP_RESPONSE_CODE_DETAIL(code));
+
+	return 0;
+}
+
+static void do_send_timeout_cb(struct lwm2m_message *msg)
+{
+	LOG_WRN("Send Timeout");
+
+}
+
+int lwm2m_engine_send(struct lwm2m_ctx *ctx, char *path_list[], uint8_t path_list_size)
+{
+	struct lwm2m_obj_path path[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
+	struct lwm2m_message *msg;
+	int ret;
+	uint16_t content_format;
+
+	if (path_list_size > CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE) {
+		return -E2BIG;
+	}
+
+	/* Select content format use CBOR when it possible */
+	if (IS_ENABLED(CONFIG_LWM2M_RW_SENML_JSON_SUPPORT)) {
+		content_format = LWM2M_FORMAT_APP_SEML_JSON;
+	} else {
+		LOG_WRN("SenML CBOR or JSON is not supported");
+		return -ENOTSUP;
+	}
+
+	/* Parse Path to internal used object path format */
+	for (int i = 0; i < path_list_size; i++) {
+		ret = string_to_path(path_list[i], path + i, '/');
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* Allocate Message buffer */
+	msg = lwm2m_get_message(ctx);
+	if (!msg) {
+		LOG_ERR("Unable to get a lwm2m message!");
+		return -ENOMEM;
+	}
+
+	msg->type = COAP_TYPE_CON;
+	msg->code = COAP_METHOD_POST;
+	msg->mid = coap_next_id();
+	msg->tkl = LWM2M_MSG_TOKEN_GENERATE_NEW;
+	msg->reply_cb = do_send_reply_cb;
+	msg->message_timeout_cb = do_send_timeout_cb;
+	msg->out.out_cpkt = &msg->cpkt;
+
+	ret = lwm2m_init_message(msg);
+	if (ret) {
+		goto cleanup;
+	}
+
+
+	ret = select_writer(&msg->out, content_format);
+	if (ret) {
+		goto cleanup;
+	}
+
+	ret = coap_packet_append_option(&msg->cpkt, COAP_OPTION_URI_PATH,
+					LWM2M_DP_CLIENT_URI,
+					strlen(LWM2M_DP_CLIENT_URI));
+	if (ret < 0) {
+		goto cleanup;
+	}
+
+	/* Write requested path data */
+	ret = do_send_op(msg, content_format, path, path_list_size);
+	if (ret < 0) {
+		LOG_ERR("error in multi-format read (err:%d)", ret);
+		goto cleanup;
+	}
+	LOG_INF("Send op to server (/dp)");
+	lwm2m_send_message(msg);
```

Comment:
I'm a little curious about this call. It sort of conflicts with the rationale in this earlier commit reworking socket interaction: https://github.com/zephyrproject-rtos/zephyr/commit/32989a38f0c8911e83d64fc898e19d29b0800f75

Not saying it's wrong to do it like this, just want to ensure that we don't end up in a similar situation that triggered the previous rework.

Reply:
OK. I can change to asynch send. No problem. When I was testing feature I see I got message out quicker by using this API.

Reply:
Yeah, it will get out faster this way. I was wondering if there's a way to get the same behavior (~instant send) using the async API, but that would require returning from `poll()` before it times out, which I don't know how to achieve.

Reply:
OK. I will change to asynch send call. Thanks for pointing this.